File: ios/Classes/Tasks/GeocodeTask.m
//
//  GeocodeTask.m
//  geolocator
//
//  Created by Maurits van Beusekom on 20/05/2019.
//

#import "GeocodeTask.h"


@implementation GeocodeTask

- (instancetype)initWithContext:(TaskContext *)context completionHandler:(CompletionHandler)completionHandler {
    self = [super initWithContext:context completionHandler:completionHandler];
    if (self) {
        //
    }
    
    return self;
}

- (void)geocodeCompletionHandlerPlacemarks:(NSArray<CLPlacemark *> *)placemarks errorCode:(NSString *)errorCode error:(NSError *)error {
    if (placemarks != nil) {
        [self processPlacemarks:placemarks];
    }
    
    if (error != nil) {
        [self handleErrorCode:errorCode message:error.localizedDescription];
    }
    
    [self stopTask];
}

- (void)processPlacemarks:(NSArray<CLPlacemark *> *)placemarks {
    NSMutableArray<NSDictionary *> *locations = [[NSMutableArray alloc] init];
    
    for (CLPlacemark *placemark in  placemarks) {
        [locations addObject:[placemark toDictionary]];
    }
    
    if (locations.count > 0) {
        [[self context] resultHandler](locations);
    } else {
        [self handleErrorCode:@"ERROR_GEOCODING_ADDRESSNOTFOUND" message:@"Could not find any result for the supplied address or coordinates."];
    }
}

- (NSString *)languageCode:(NSLocale *)locale {
    if (@available(iOS 10.0, *)) {
        return [locale languageCode];
    } else {
        return [[locale localeIdentifier] substringToIndex:2];
    }
}

@end


@implementation ForwardGeocodeTask {
    NSString *_address;
    NSLocale *_locale;
}

- (instancetype)initWithContext:(TaskContext *)context completionHandler:(CompletionHandler)completionHandler {
    self = [super initWithContext:context completionHandler:completionHandler];
    if (self) {
        //
    }
    
    [self parseArguments:context.arguments];
    return self;
}

- (void)parseArguments:(id)arguments {
    if ([arguments isKindOfClass:[NSDictionary class]]) {
        _address = arguments[@"address"];
        
        if (arguments[@"localeIdentifier"] != nil) {
            _locale = [[NSLocale alloc] initWithLocaleIdentifier:arguments[@"localeIdentifier"]];
        }
    } else {
        _address = nil;
    }
}

- (void)startTask {
    if (_address == nil) {
        [self handleErrorCode:@"NO_ADDRESS_SUPPLIED" message:@"Please supply a valid string containing the address to lookup"];
        
        [self stopTask];
        return;
    }
    
    [self geocodeAddress:_address];
}

- (void)geocodeAddress:(NSString *)address {
    CLGeocoder *geocoding = [[CLGeocoder alloc] init];
    
    if (@available(iOS 11.0, *)) {
        [geocoding geocodeAddressString:address inRegion:nil preferredLocale:_locale completionHandler:^(NSArray< CLPlacemark *> *__nullable placemarks, NSError *__nullable error) {
            [self geocodeCompletionHandlerPlacemarks:placemarks errorCode:@"ERROR_GEOCODING_ADDRESS" error:error];
        }];
    } else {
        NSUserDefaults *standardUserDefaults = [NSUserDefaults standardUserDefaults];
        NSArray<NSString * > *defaultLanguages;
        
        if (_locale != nil) {
            defaultLanguages = [standardUserDefaults arrayForKey:@"AppleLanguages"];
            [standardUserDefaults setValue:[self languageCode:_locale] forKey:@"AppleLanguages"];
        }
        
        [geocoding geocodeAddressString:address completionHandler:^(NSArray< CLPlacemark *> *__nullable placemarks, NSError *__nullable error) {
            [self geocodeCompletionHandlerPlacemarks:placemarks errorCode:@"ERROR_GEOCODING_ADDRESS" error:error];
            
            if (self->_locale != nil) {
                [standardUserDefaults setValue:defaultLanguages forKey:@"AppleLanguages"];
            }
        }];
    }
}

@end


@implementation ReverseGeocodeTask {
    CLLocation *_location;
    NSLocale *_locale;
}

- (instancetype)initWithContext:(TaskContext *)context completionHandler:(CompletionHandler)completionHandler {
    self = [super initWithContext:context completionHandler:completionHandler];
    if (self) {
        //
    }
    
    [self parseArguments:context.arguments];
    return self;
}

- (void)parseArguments:(id)arguments {
    if ([arguments isKindOfClass:[NSDictionary class]]) {
        CLLocationDegrees latitude = ((NSNumber *) arguments[@"latitude"]).doubleValue;
        CLLocationDegrees longitude = ((NSNumber *) arguments[@"longitude"]).doubleValue;
        
        _location = [[CLLocation alloc] initWithLatitude:latitude longitude:longitude];
        
        if (arguments[@"localeIdentifier"] != nil) {
            _locale = [[NSLocale alloc] initWithLocaleIdentifier:(NSString *) arguments[@"localeIdentifier"]];
        }
    } else {
        _location = nil;
    }
}

- (void)startTask {
    if (_location == nil) {
        [self handleErrorCode:@"NO_COORDINATES_SUPPLIED" message:@"Please supply valid latitude and longitude values."];
        [self stopTask];
        return;
    }
    
    
    [self reverseToAddress:_location];
}

- (void)reverseToAddress:(CLLocation *)location {
    CLGeocoder *geocoding = [[CLGeocoder alloc] init];
    
    if (@available(iOS 11.0, *)) {
        [geocoding reverseGeocodeLocation:location preferredLocale:_locale completionHandler:^(NSArray< CLPlacemark *> *__nullable placemarks, NSError *__nullable error) {
            [self geocodeCompletionHandlerPlacemarks:placemarks errorCode:@"ERROR_REVERSEGEOCODING_LOCATION" error:error];
        }];
    } else {
        NSUserDefaults *standardUserDefaults = [NSUserDefaults standardUserDefaults];
        NSArray<NSString * > *defaultLanguages;
        
        if (_locale != nil) {
            defaultLanguages = [standardUserDefaults arrayForKey:@"AppleLanguages"];
            [standardUserDefaults setValue:[self languageCode:_locale] forKey:@"AppleLanguages"];
        }
        
        [geocoding reverseGeocodeLocation:location completionHandler:^(NSArray< CLPlacemark *> *__nullable placemarks, NSError *__nullable error) {
            [self geocodeCompletionHandlerPlacemarks:placemarks errorCode:@"ERROR_REVERSEGEOCODING_LOCATION" error:error];
            
            if (self->_locale != nil) {
                [standardUserDefaults setValue:defaultLanguages forKey:@"AppleLanguages"];
            }
        }];
    }
}

@end

File: ios/Classes/Tasks/GeocodeTask.swift
//
//  GeocodingService.swift
//  geolocator
//
//  Created by Maurits van Beusekom on 03/07/2018.
//

import CoreLocation
import Foundation

class GeocodeTask: Task {   
    required init(context: TaskContext,
         completionHandler: CompletionHandler?) {

        super.init(context: context,
                   completionHandler: completionHandler)
    }
    
    func geocodeCompletionHandler(placemarks: [CLPlacemark]?, errorCode: String, error: Error?) {
        if let marks = placemarks {
            self.processPlacemark(placemarks: marks);
        }
        
        if let err = error {
            self.handleError(
                code: errorCode,
                message: err.localizedDescription)
        }
        
        self.stopTask();
    }
    
    func processPlacemark(placemarks: [CLPlacemark]) {
        var locations: [NSDictionary] = []
        for placemark in placemarks {
            locations.append(placemark.toDictionary())
        }
        
        if locations.count > 0 {
            context.resultHandler(locations)
        } else {
            handleError(
                code: "ERROR_GEOCODING_ADDRESSNOTFOUND",
                message: "Could not find any result for the supplied address or coordinates.")
        }
    }
}

class ForwardGeocodeTask: GeocodeTask, TaskProtocol {
    private var _address: String?
    private var _locale: Locale?
    
    required init(context: TaskContext,
                  completionHandler: CompletionHandler?) {

        super.init(context: context,
                   completionHandler: completionHandler)
        
        parseArguments(arguments: context.arguments)
    }
    
    func parseArguments(arguments: Any?) {
        if let argumentMap = context.arguments as! NSDictionary? {
            _address = argumentMap["address"] as? String
            
            if let localeIdentifier = argumentMap["localeIdentifier"] {
                _locale = Locale(identifier: localeIdentifier as! String)
            }
        } else {
            _address = nil
        }
    }
    
    func startTask() {
        guard let address = _address else
        {
            handleError(
                code: "NO_ADDRESS_SUPPLIED",
                message: "Please supply a valid string containing the address to lookup")
            
            stopTask()
            return
        }
        
        geocodeAddress(address: address)
    }
    
    private func geocodeAddress(address: String) {
        
        let geocoding = CLGeocoder.init()
        if #available(iOS 11.0, *) {
            geocoding.geocodeAddressString(address, in: nil, preferredLocale: _locale) {(placemarks, error) in
                self.geocodeCompletionHandler(placemarks: placemarks, errorCode: "ERROR_GEOCODING_ADDRESS", error: error)
            }
        } else {
            var defaultLanguages: [String]?
            if let locale = _locale {
                defaultLanguages = UserDefaults.standard.array(forKey: "AppleLanguages") as? [String]
                UserDefaults.standard.set([locale.languageCode], forKey: "AppleLanguages")
            }
            
            geocoding.geocodeAddressString(address) { (placemarks, error) in
                self.geocodeCompletionHandler(placemarks: placemarks, errorCode: "ERROR_GEOCODING_ADDRESS", error: error)
                
                if self._locale != nil {
                    UserDefaults.standard.set(defaultLanguages, forKey: "AppleLanguages")
                }
            }
        }
    }
}

class ReverseGeocodeTask: GeocodeTask, TaskProtocol {
    private var _location: CLLocation?
    private var _locale: Locale?
    
    required init(context: TaskContext,
                  completionHandler: CompletionHandler?) {
        
        super.init(context: context,
                   completionHandler: completionHandler)
        
        parseArguments(arguments: context.arguments)
    }
    
    func parseArguments(arguments: Any?) {
        if let argumentMap = context.arguments as! NSDictionary? {
            let latitude = argumentMap["latitude"] as! CLLocationDegrees
            let longitude = argumentMap["longitude"] as! CLLocationDegrees
            
            _location = CLLocation.init(latitude: latitude, longitude: longitude)
            
            if let localeIdentifier = argumentMap["localeIdentifier"] {
                _locale = Locale(identifier: localeIdentifier as! String)
            }
        } else {
            _location = nil
        }
    }
    
    func startTask() {
        guard let location = _location else
        {
            handleError(
                code: "NO_COORDINATES_SUPPLIED",
                message: "Please supply valid latitude and longitude values.")
            
            stopTask()
            return
        }
        
        reverseToAddress(location: location)
    }
    
    private func reverseToAddress(location: CLLocation) {
        
        let geocoding = CLGeocoder.init()
        
        if #available(iOS 11.0, *) {
            geocoding.reverseGeocodeLocation(location, preferredLocale: _locale) { (placemarks, error) in
                self.geocodeCompletionHandler(placemarks: placemarks, errorCode: "ERROR_REVERSEGEOCODING_LOCATION", error: error)
            }
        } else {
            var defaultLanguages: [String]?
            if let locale = _locale {
                defaultLanguages = UserDefaults.standard.array(forKey: "AppleLanguages") as? [String]
                UserDefaults.standard.set([locale.languageCode], forKey: "AppleLanguages")
            }
            
            geocoding.reverseGeocodeLocation(location)  { (placemarks, error) in
                self.geocodeCompletionHandler(placemarks: placemarks, errorCode: "ERROR_REVERSEGEOCODING_LOCATION", error: error)
                
                if self._locale != nil {
                    UserDefaults.standard.set(defaultLanguages, forKey: "AppleLanguages")
                }
            }
        }
    }
}
